src/lib.rs

```diff
@@ -117,17 +118,120 @@ pub fn select_coin_knapsack(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut adjusted_target = options.target_value
+        + options.min_drain_value
+        + calculate_fee(options.base_weight, options.target_feerate);
+    let mut smaller_coins = inputs
+        .iter()
+        .enumerate()
+        .filter(|&(index, output_group)| output_group.value < adjusted_target)
+        .map(|(index, output_group)| (index, *output_group))
+        .collect::<Vec<_>>();
+    // Sorting smaller_coins in descending order
+    smaller_coins.sort_by_key(|&(_, output_group)| Reverse(output_group.value));
+
+    knap_sack(adjusted_target, &smaller_coins, inputs, options)
 }
 
 /// adjusted_target should be target value plus estimated fee
 /// smaller_coins is a slice of pair where the usize refers to the index of the OutputGroup in the inputs given
 /// smaller_coins should be sorted in descending order based on the value of the OutputGroup, and every OutputGroup value should be less than adjusted_target
+fn knap_sack_calculate_accumulated_weight(
```

Comment:
`calculate_accumulated_weight` should be fine.

`selected_inputs` could be a reference to an Iterator rather than a HashSet, so in future this can be used with `Vec<usize>` as well.